### PR TITLE
[UNR-5131] Remove the authoritative destruction switch

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Remove the authoritative destruction switch, as it should no longer be required in the world of USLB.
This should also allow new ways of destroying actors from game code.

#### Release note
N/A?

#### Tests
Ran the test suite. Hm, could do something more.

#### Documentation
N/A

#### Primary reviewers
idk

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -281,7 +276,7 @@ void USpatialActorChannel::RetireEntityIfAuthoritative()
 				NetDriver->ActorSystem->RetireEntity(EntityId, Actor->IsNetStartupActor());
 			}
 		}
-		else if (bCreatedEntity) // We have not gained authority yet
+		else if (bCreatingNewEntity) // We have not gained authority yet
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:279]: This one was fun
I think this change is correct, since we were checking for entities that should have had a CreateEntity request in flight. But I am slightly confused about the flows here, so could talk with someone about it.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -283,7 +278,7 @@ void USpatialActorChannel::RetireEntityIfAuthoritative()
 				NetDriver->ActorSystem->RetireEntity(EntityId, Actor->IsNetStartupActor());
 			}
 		}
-		else if (bCreatedEntity) // We have not gained authority yet
+		else if (NetDriver->ActorSystem->IsCreateEntityRequestInFlight(EntityId)) // We have not gained authority yet
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:281]: Turns out I can't check the view here to see if the entity doesn't exist locally, because we are responding to a removeEntityOp here, so the view would have already been updated with the entity deleted.
I decided to implement part of @samiwh's recommendation to have a list of createEntityRequests in flight, but didn't see a good way to do it at the view level, so decided to do it at the ActorSystem level.

Open to any suggestions here - as it stands currently, it's theoretically more efficient to have a flag on the actor channel conveying this information.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 3]
```diff
@@ -715,14 +710,6 @@ int64 USpatialActorChannel::ReplicateActor()
...
 
 			SetAutonomousProxyOnAuthority(Actor->RemoteRole == ROLE_AutonomousProxy);
-
-			// We preemptively set the Actor role to SimulatedProxy if load balancing is disabled
-			// (since the legacy behaviour is to wait until Spatial tells us we have authority)
-			if (NetDriver->LoadBalanceStrategy == nullptr)
-			{
-				Actor->Role = ROLE_SimulatedProxy;
-				Actor->RemoteRole = ROLE_Authority;
-			}
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:725]: drive-by removal, as LBStrategy is now always non-null

could add an ensure somewhere or something...

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h [hunk 4]
```diff
@@ -206,6 +207,9 @@ class ActorSystem
 
 	TSet<Worker_EntityId_Key> PresentEntities;
 
+	// Only the ones we know the EntityId for beforehand.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:210]: When isn't this the case?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 5]
```diff
@@ -2409,6 +2418,7 @@ void ActorSystem::OnEntityCreated(const Worker_CreateEntityResponseOp& Op, FSpat
 
 	AActor* Actor = Channel.Actor;
 	const Worker_EntityId EntityId = Channel.GetEntityId();
+	ensure(EntityId != SpatialConstants::INVALID_ENTITY_ID);
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2421]: New ensure based on Mike's comment about how the entity pool works where we should always be creating entities for actors with known entity ids.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 6]
```diff
@@ -2438,6 +2448,8 @@ void ActorSystem::OnEntityCreated(const Worker_CreateEntityResponseOp& Op, FSpat
 			   TEXT("Create entity request succeeded. "
 					"Actor %s, request id: %d, entity id: %lld, message: %s"),
 			   *Actor->GetName(), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+		ensure(EntityId == Op.entity_id);
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2451]: also thought I would throw in the expectation that the request has to be for the same entity id as the actor channel currently has

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h (hunk 4 above)
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:210]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: it was not correct

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h]: Sorry, removed. Leftover from my early ignorance.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 7]
```diff
@@ -2347,6 +2347,10 @@ void ActorSystem::SendCreateEntityRequest(USpatialActorChannel& ActorChannel, ui
 			OnEntityCreated(Op, SpanId);
 		});
 
+		if (ensure(EntityId != SpatialConstants::INVALID_ENTITY_ID))
+		{
+			CreateEntityRequestsInFlight.Add(EntityId);
+		}
 		CreateEntityRequestIdToActorChannel.Emplace(CreateEntityRequestId, MakeWeakObjectPtr(&ActorChannel));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2354]: `CreateEntityRequestIdToActorChannel` actually seems redundant now, as we can get the actor from the entity id, and we don't need the actor channel at all

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2354]: But that's not a cleanup required now

**Review by m-samiec — APPROVED** [review]
Nice work

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:281]: This should be at the actor system level. Although rather than it be actual system command requests, I'd make it more abstractly - what is the worker currently creating. The start point is when you decide you want to create a new entity and the end point is when you receive authority over that entity. The actual create request itself starts after that and ends before that.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 8]
```diff
@@ -2390,8 +2394,15 @@ bool ActorSystem::HasPendingOpsForChannel(const USpatialActorChannel& ActorChann
...
 }
 
+bool ActorSystem::IsCreateEntityRequestInFlight(Worker_EntityId EntityId) const
+{
+	return CreateEntityRequestsInFlight.Contains(EntityId);
+}
+
 void ActorSystem::OnEntityCreated(const Worker_CreateEntityResponseOp& Op, FSpatialGDKSpanId CreateOpSpan)
 {
+	CreateEntityRequestsInFlight.Remove(Op.entity_id);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2404]: To my other comment - this should be on authority gain. This might come on the same tick or might not.
